graylog2-server/src/main/java/org/graylog2/bootstrap/preflight/PreflightConfigServiceImpl.java

```diff
@@ -16,36 +16,54 @@
  */
 package org.graylog2.bootstrap.preflight;
 
-import com.google.common.collect.ImmutableMap;
 import com.mongodb.BasicDBObject;
+import com.mongodb.DBCollection;
 import com.mongodb.DBObject;
-import org.bson.types.ObjectId;
 import org.graylog2.database.MongoConnection;
-import org.graylog2.database.PersistedServiceImpl;
-import org.graylog2.plugin.database.ValidationException;
 
 import javax.inject.Inject;
 import java.util.Optional;
 
-public class PreflightConfigServiceImpl extends PersistedServiceImpl implements PreflightConfigService {
+public class PreflightConfigServiceImpl implements PreflightConfigService {
+
+    public static final String COLLECTION_NAME = "preflight";
+
+    private final MongoConnection connection;
 
     @Inject
     public PreflightConfigServiceImpl(MongoConnection connection) {
-        super(connection);
+        this.connection = connection;
+    }
+
+    @Override
+    public PreflightConfig setConfigResult(PreflightConfigResult result) {
+        getCollection()
+                .update(new BasicDBObject("type", "preflight_result"),
+                        new BasicDBObject("$set", new BasicDBObject("value", result)),
+                        true,
+                        false
+                );
+        return new PreflightConfig(result);
     }
 
     @Override
-    public Optional<PreflightConfig> getPersistedConfig() {
-        final DBObject doc = findOne(PreflightConfigImpl.class, new BasicDBObject());
+    public PreflightConfigResult getPreflightConfigResult() {
+        final DBObject doc = getCollection().findOne(new BasicDBObject("type", "preflight_result"));
         return Optional.ofNullable(doc)
-                .map(o -> new PreflightConfigImpl((ObjectId) o.get("_id"), o.toMap()));
+                .map(d -> (String) d.get("value"))
+                .map(PreflightConfigResult::valueOf)
+                .orElse(PreflightConfigResult.UNKNOWN);
+    }
+
+    private DBCollection getCollection() {
+        return this.connection.getDatabase().getCollection(COLLECTION_NAME);
     }
 
     @Override
-    public PreflightConfig saveConfiguration() throws ValidationException {
-        final ImmutableMap<String, Object> fields = ImmutableMap.of("result", PreflightConfigResult.FINISHED);
-        final PreflightConfigImpl config = new PreflightConfigImpl(fields);
-        final String id = save(config);
-        return getPersistedConfig().orElseThrow(() -> new IllegalStateException("Failed to obtain configuration that was just stored"));
+    public String getPreflightPassword() {
+        final DBObject doc = getCollection().findOne(new BasicDBObject("type", "preflight_password"));
+        return Optional.ofNullable(doc)
+                .map(d -> (String) d.get("value"))
+                .orElseThrow(() -> new IllegalStateException("Initial password should be present in the DB, inconsistent state"));
```

Comment:
I just wander what a user can do after receiving this message? Maybe it should tell what kind of document should be inserted manually to MongoDB to make it run?

Reply:
Yeah, I am unhappy with the message as well. It should never happen and if that happens it's completely our fault, the user can't do anything about it. So we should make that clear and tell them to contact us, I believe.

Reply:
I'm not sure. At this point, isn't it the case that nothing is really configured at this point, so starting over and dropping the mongo db would be sensible?